File: 자료구조_큐, 우선순위 큐/프로그래머스lv2-스킬트리.py
from collections import deque, defaultdict
def check_skill(key:str, target:str)->int:
    dic = defaultdict(int)
    for i, v in enumerate(key):
        dic[v] = i + 1
    visited = [True] + [False] * 26
    q = deque()
    for a in target:
        q.append(a)
    
    while q:
        x = q.popleft()
        if dic[x]:
            if visited[dic[x] - 1]:
                visited[dic[x]] = True
            else:
                return 0
    if not q:
        return 1


def solution(skill, skill_trees):
    result = 0
    for skill_tree in skill_trees:
        result += check_skill(skill, skill_tree)
    
    return result

"""
정확성  테스트
테스트 1 〉	통과 (0.02ms, 10.2MB)
테스트 2 〉	통과 (0.02ms, 10.2MB)
테스트 3 〉	통과 (0.02ms, 10.3MB)
테스트 4 〉	통과 (0.02ms, 10.2MB)
테스트 5 〉	통과 (0.03ms, 10.2MB)
테스트 6 〉	통과 (0.03ms, 10.2MB)
테스트 7 〉	통과 (0.04ms, 10.3MB)
테스트 8 〉	통과 (0.03ms, 10.2MB)
테스트 9 〉	통과 (0.03ms, 10.2MB)
테스트 10 〉	통과 (0.02ms, 10.3MB)
테스트 11 〉	통과 (0.03ms, 10.3MB)
테스트 12 〉	통과 (0.05ms, 10.3MB)
테스트 13 〉	통과 (0.04ms, 10.2MB)
테스트 14 〉	통과 (0.03ms, 10.3MB)
"""

# 개선에 개선을 거듭한 풀이 (deque.copy()로 아예 처음부터 복사본을 넘겨줌)

from collections import deque
# 처음부터 skill 큐를 인자로 받는다.
def check_skill(skill:deque, skill_tree:str)->int:
    # 따라서 제시된 스킬트리만 큐로 생성한다.
    q = deque()
    for s in skill_tree:
        if s in skill:
            q.append(s)
    while skill and q:
        if skill.popleft() != q.popleft():
            return 0
    return 1

def solution(skill, skill_trees):
    # 선행스킬에 대한 큐
    q = deque()
    for s in skill:
        q.append(s)

    result = 0
    for skill_tree in skill_trees:
        # deque.copy()를 이용하여 복사본을 넘겨줌
        result += check_skill(q.copy(), skill_tree)
    return result

"""
정확성  테스트
테스트 1 〉	통과 (0.02ms, 10.2MB)
테스트 2 〉	통과 (0.02ms, 10.3MB)
테스트 3 〉	통과 (0.01ms, 10.3MB)
테스트 4 〉	통과 (0.02ms, 10.2MB)
테스트 5 〉	통과 (0.02ms, 10.2MB)
테스트 6 〉	통과 (0.01ms, 10.2MB)
테스트 7 〉	통과 (0.03ms, 10.2MB)
테스트 8 〉	통과 (0.02ms, 10.2MB)
테스트 9 〉	통과 (0.02ms, 10.3MB)
테스트 10 〉	통과 (0.01ms, 10.2MB)
테스트 11 〉	통과 (0.02ms, 10.2MB)
테스트 12 〉	통과 (0.03ms, 10.3MB)
테스트 13 〉	통과 (0.02ms, 10.2MB)
테스트 14 〉	통과 (0.01ms, 10.3MB)
"""

File: 큐/프로그래머스lv2-스킬트리.js
function checkSkill(skill, skillTrees) {
  skill.reverse()
  skillTrees.reverse()
  while (skill.length > 0 && skillTrees.length > 0) {
    if (skill.pop() !== skillTrees.pop()) return 0
  }
  return 1
}

function solution(skill, skill_trees) {
  const dict = {}
  for (let i = 0; i < skill.length; i++) {
    dict[skill[i]] = i + 1
  }

  let result = 0

  for (const skillTree of skill_trees) {
    let shortSkill = []
    for (const char of skillTree) {
      if (dict[char]) shortSkill.push(char)
    }
    result += checkSkill([...skill], shortSkill)
  }
  return result
}

// 정확성  테스트
// 테스트 1 〉	통과 (0.14ms, 30.1MB)
// 테스트 2 〉	통과 (0.14ms, 29.9MB)
// 테스트 3 〉	통과 (0.20ms, 30.2MB)
// 테스트 4 〉	통과 (0.17ms, 30.1MB)
// 테스트 5 〉	통과 (0.15ms, 30.2MB)
// 테스트 6 〉	통과 (0.14ms, 30MB)
// 테스트 7 〉	통과 (0.42ms, 29.9MB)
// 테스트 8 〉	통과 (0.16ms, 30.1MB)
// 테스트 9 〉	통과 (0.34ms, 30.1MB)
// 테스트 10 〉	통과 (0.14ms, 30.2MB)
// 테스트 11 〉	통과 (0.43ms, 30.1MB)
// 테스트 12 〉	통과 (0.46ms, 30MB)
// 테스트 13 〉	통과 (0.16ms, 30.2MB)
// 테스트 14 〉	통과 (0.14ms, 30.2MB)
